tests/solana_tests/abi_decode.rs

```diff
@@ -0,0 +1,1175 @@
+        vec_1: Vec<S>,
+        vec_2: Vec<String>,
+    }
+
+    let mut vm = build_solidity(
+        r#"
+    contract Testing {
+        struct S {
+            int64 f1;
+            string f2;
+        }
+
+        function testEmpty(bytes memory buffer) public pure {
+            (S[] memory vec_1, string[] memory vec_2) = abi.borshDecode(buffer, (S[], string[]));
+
+            assert(vec_1.length == 0);
+            assert(vec_2.length == 0);
+        }
+    }
+        "#,
+    );
+    vm.constructor("Testing", &[]);
+
+    let input = Input {
+        vec_1: vec![],
+        vec_2: vec![],
+    };
+    let encoded = input.try_to_vec().unwrap();
+    let _ = vm.function("testEmpty", &[Token::Bytes(encoded)], &[], None);
+}
+
+#[test]
+fn external_function() {
+    #[derive(Debug, BorshSerialize)]
+    struct Input {
+        selector: [u8; 4],
+        address: [u8; 32],
+    }
+
+    let mut vm = build_solidity(
+        r#"
+    contract Testing {
+        function testExternalFunction(bytes memory buffer) public view returns (bytes4, address) {
+            function (uint8) external returns (int8) fPtr = abi.borshDecode(buffer, (function (uint8) external returns (int8)));
+            return (fPtr.selector, fPtr.address);
+        }
+    }
+        "#,
+    );
+
+    vm.constructor("Testing", &[]);
+    let input = Input {
+        selector: [1, 2, 3, 4],
+        address: [
+            0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23,
+            24, 25, 26, 27, 28, 29, 30, 31,
+        ],
+    };
+    let encoded = input.try_to_vec().unwrap();
+    let returns = vm.function("testExternalFunction", &[Token::Bytes(encoded)], &[], None);
+
+    let selector = returns[0].clone().into_fixed_bytes().unwrap();
+    assert_eq!(selector, input.selector);
+
+    let address = returns[1].clone().into_fixed_bytes().unwrap();
+    assert_eq!(address, input.address);
+}
+
+#[test]
+fn bytes_arrays() {
+    #[derive(Debug, BorshSerialize)]
+    struct Input {
+        item_1: [[u8; 4]; 2],
+        item_2: Vec<[u8; 5]>,
+    }
+
+    let mut vm = build_solidity(
+        r#"
+        contract Testing {
+            function testByteArrays(bytes memory buffer) public view {
+                (bytes4[2] memory arr, bytes5[] memory vec) = abi.borshDecode(buffer, (bytes4[2], bytes5[]));
+
+                assert(arr[0] == "abcd");
+                assert(arr[1] == "efgh");
+
+                assert(vec.length == 2);
+                assert(vec[0] == "12345");
+                assert(vec[1] == "67890");
+            }
+        }
+        "#,
+    );
+
+    vm.constructor("Testing", &[]);
+    let input = Input {
+        item_1: [b"abcd".to_owned(), b"efgh".to_owned()],
+        item_2: vec![b"12345".to_owned(), b"67890".to_owned()],
+    };
+    let encoded = input.try_to_vec().unwrap();
+    let _ = vm.function("testByteArrays", &[Token::Bytes(encoded)], &[], None);
+}
+
+#[test]
+#[should_panic]
```

Comment:
expected = ".."

Reply:
I really wanted to use this Rust feature, but the eBPF VM returns an error code, which makes the following assertion fail. The code panics without any error message, so I cannot use `#[should_panic(expected = "my_error")]`. This is why I suggested that the VM should return the error code, instead of panicing.

https://github.com/hyperledger-labs/solang/blob/b867c8a6c7a1ee89d405993abef458fc59e9b0fb/tests/solana.rs#L1583

Reply:
That `assert_eq!()` should be re-written. Does:
```
assert_eq!(res.unwrap(), 0));
```
give you the correct error message? (See https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c9db9b9df941e8bde46f7d3a9db72aa0)

Reply:
The error message I got now is
```
assertion failed: `(left == right)`
  left: `Ok(4294967296)`,
 right: `Ok(0)`
```
If I change to your suggestion, it turns to:
```
assertion failed: `(left == right)`
  left: `4294967296`,
 right: `0`
```
Keep in mind that this solution does not account for when the VM returns Err (not Ok()).

One solution I found now was adding this decorator to the test function:
```
#[should_panic(expected = "assertion failed: `(left == right)`
  left: `Ok(4294967296)`,
 right: `Ok(0)`")]
```

Reply:
That number makes more sense when written in hex. Maybe it should do something like:
```
match res {
    Ok(0) => (),
    Ok(v) =>panic!("unexpected return {:#x}", v),
    Err(e) => panic!("error: {}", e);
}
```

Reply:
Good idea